### ISSUE TITLE
repeat command ?

### ISSUE THREAD

**Issue by gautaz** [webdriverio/webdriverio#636]
Hello,

I am wondering if there was a simple way to instruct a webdriverio client to repeat an operation.

Currently, I have come up with the following solution:
```
client = webdriverio.remote(options);

client.addCommand('repeat', function repeat(times, operation) {                                                                                             
  if (times > 0) {
    --arguments[0];                                                                                                                                      
    return repeat.apply(                                                                                                                                 
      this[operation].apply(this, Array.prototype.slice.call(arguments, 2)),                                                                           
      arguments                                                                                                                                        
    );                                                                                                                                                   
  } else {                                                                                                                                                 
    return this;                                                                                                                                         
  }                                                                                                                                                        
});    
```

This way, I can further write sentences like :
```
client.url('http://yada.com').repeat(15, 'click', 'selector')
```

I have no idea if there is a better way to do this with the current webdriverio API, did I miss something ?

**Comment by christian-bromann** [comment]
You could also use `waitUntil` for that matter. It repeats everything that is within the callback function until a certain timeout, so for example:

```js
var times = 15;
var i = 0;
client.waitUntil(function() {
    return client.click('#someSelector').then(function() {
        i++;
        return i === 15;
    });
}, 10000);
```
It might has the advantage that it stops after a certain timeout (no infinite loop). Another way would be a normal loop (like for or while) but then you would kinda do everything in parallel which seems not to be desired here.

What exactly do you want to achieve here?

**Comment by gautaz** [comment]
Hello Christian, my main goal was to increase the readability of my test scenarios.

I'm currently using [promises transfer](http://webdriver.io/guide/usage/transferpromises.html) and I am trying to keep my test sentences as simple as possible.
I know that brevity is not a goal per se but it eases test writing and reading.

I admit using the *repeat* custom command has no meaning if only used once, but when it's repeated :-) all over the test suite, I found it more readable than the *waitUntil* construct.
I guess it could be improved to handle generic predicates (instead of just the *times* argument).

To broaden a bit my expectations, I think offering a way to express control structures (if / while / for / whatever...) while keeping the syntax offered by [Transfer Promises](http://webdriver.io/guide/usage/transferpromises.html) in a simple manner would somehow improve test case writing and reading.

**Comment by christian-bromann** [comment]
> increase the readability of my test scenarios

How does this increase readability in your test cases?

> I found it more readable than the waitUntil construct.

I agree.

> To broaden a bit my expectations, I think offering a way to express control structures (if / while / for / whatever...) while keeping the syntax offered by Transfer Promises in a simple manner would somehow improve test case writing and reading.

Actually you can use all those things (unfortunately not yet in Cucumber). In Mocha or Jasmine you can make use of generators and can do things like:

```js
it('should prompt a nice hello message, function*() {
    yield browser.url('http://mywebsite.com');

    if(yield browser.getText(selector) === 'Mr.') {
        yield browser.setValue(selector2, 'Hello Sir!');
    } else {
        yield browser.setValue(selector2, 'Hello Madam');
    }

    for(var i = 0; i < 15; i++) {
        yield browser.click(selector3);
    }
});
```

**Comment by gautaz** [comment]
> How does this increase readability in your test cases?

Here is just an excerpt of one of my test cases:
```javascript
it('removes 40 items and ensures both previous and next are disabled', function () {
  this.timeout(10000);
  return browse({ pathname: 'admin' })
  .repeat(40, 'click', '.item-description button.delete')
  .getAttribute('#previous-page', 'disabled').should.eventually.equal('true')
  .getAttribute('#next-page', 'disabled').should.eventually.equal('true');
});
```

> Actually you can use all those things (unfortunately not yet in Cucumber). In Mocha or Jasmine you can make use of generators and can do things like:

That is quite close to the syntax I am looking for :-).
Unfortunately, I'm currently constrained to stick to node.js 0.10.
AFAIK, *yield* is ecmascript 6, am I right ?

**Comment by christian-bromann** [comment]
yeah but available with node `v.0.11.x`
Ok I guess we can close this issue or is there anything I missed to answer?

**Comment by gautaz** [comment]
Thank you very much for your time and your answers, I'm closing the issue right now.
I will push as hard as I can so I can use node 0.12, coming from python, there is so much features I'm missing like generators and co ;-).

**Comment by captainalabi** [comment]
i am testing an ecommerce site which involves clicking an increment button to set the quantity to buy before adding to cart. I have to repeat this sequence for 4 products before clicking the cart button. I have implemented the clicking of the button with a for loop but the result is not consistent with what I expect. The url is https://rahulshettyacademy.com/seleniumPractise/#/cart.


async multiIncrement(amount){
    let i;
    for(i = 1; i <= amount; i++){
                await this.clickIncrementBtn()
             }
       i=0;
}
What i get is that when I run the code with 3, the button is clicked 4 times and the subsequent run of the code with 2 clicks the button 5 times.
How can i make the result of the for loop to be consistent

**Comment by christian-bromann** [comment]
Join our  [Matrix](https://matrix.to/#/#webdriver.io:gitter.im) support chat for these kind of questions. This issue is over 7 years old.

**Comment by captainalabi** [comment]
Thanks for your response to my question. I'll view it on GitHub.

Alabi Razaq

**Comment by cg-tester** [comment]
@christian-bromann  how can I use client object ?  and in async testing,  how can I use client.waitUntil approach you mentioned ?

**Comment by christian-bromann** [comment]
> how can I use client object ?

What do you mean by that?

> how can I use client.waitUntil approach you mentioned ?

Please see the docs https://webdriver.io/docs/api/browser/waitUntil

**Comment by cg-tester** [comment]
@christian-bromann 
I mean, your comment in July 16 2015 above.
first, in my side, client.waitUntil() is not a function error occurs.
second, you mentioned in sync way. and how to in async way ?

**Comment by cg-tester** [comment]
@christian-bromann  
I want to repeat to click next button until $('selector').getText() is equal to "mytext".
Could you help me ?

**Comment by christian-bromann** [comment]
@cg-tester not sure what your setup looks like, but if you using latest WebdriverIO testrunner, this should work:

```ts
import { browser } from '@wdio/globals'

describe('foo'. () => {
  it('bar', () => {
    await browser.waitUntil(async () => {
      const text = await $('...').getText()
      return text === '...'
    })
  })
})

**Comment by cg-tester** [comment]
@christian-bromann   
sorry but, where is click action ?

**Comment by christian-bromann** [comment]
> sorry but, where is click action ?

You can call click commands inside the waitUntil or outside.

**Comment by cg-tester** [comment]
@christian-bromann 
it works well.  thank you very much.  what can I do for you ?

**Comment by christian-bromann** [comment]
> it works well. thank you very much. what can I do for you ?

Awesome! Join our [Discord](https://discord.webdriver.io) server in case you have further questions. Happy Testing!

**Comment by cg-tester** [comment]
@christian-bromann 
can I work with you ?

**Comment by christian-bromann** [comment]
> can I work with you ?

Work with me on what?

**Comment by cg-tester** [comment]
@christian-bromann 
I like anything special that can be solved by using brain.

**Comment by christian-bromann** [comment]
Gotcha, well, we are always looking for more contributors on the WebdriverIO project. So any contributions are welcome!

**Comment by cg-tester** [comment]
@christian-bromann  
so what can I do ?

**Comment by christian-bromann** [comment]
Take a look here: https://webdriver.io/docs/contribute

**Comment by cg-tester** [comment]
@christian-bromann 
I have had a look. could you tell me more about what to do and can we meet on skype ?

**Comment by christian-bromann** [comment]
> I have had a look. could you tell me more about what to do and can we meet on skype ?

I recommend to look at all the [open issues](https://github.com/webdriverio/webdriverio/issues), especially the ones labeled with ["good first pick"](https://github.com/webdriverio/webdriverio/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+pick%22). You can request a [pairing session](https://webdriver.io/community/openofficehours) once you've committed to a specific bug you like to fix.

**Comment by cg-tester** [comment]
@christian-bromann 
I only see list of issues and comments.
How can I enter a project that I can investigate and fix bugs ?

**Comment by christian-bromann** [comment]
> How can I enter a project that I can investigate and fix bugs ?

I am not sure what you mean. I recommend to join our  [Discord](https://discord.webdriver.io) server for these kind of questions.

**Comment by cg-tester** [comment]
@christian-bromann 
now, I can't enter browser. I mean,
I command ' npm run wdio ',  the browser opens and immediately close automatically.  of course it works well before. 
I can't find the reason.  Could you tell me the reason ?
